### PR TITLE
Various s390 plugins fixes

### LINKED ISSUE #274: Fails to build on s390x: #include <s390utils/vtoc.h>
libblock 2.12
Debian unstable

libblockdev fails to build on s390x.

https://github.com/storaged-project/libblockdev/blob/master/src/lib/plugin_apis/s390.api#L7

Where is `s390utils/vtoc.h` ?

[Full build log](https://buildd.debian.org/status/fetch.php?pkg=libblockdev&arch=s390x&ver=2.12-2&stamp=1504190590)

Build log excerpt
---------------------
```
make[4]: Leaving directory '/<<PKGBUILDDIR>>/src/lib/plugin_apis'
make[4]: Entering directory '/<<PKGBUILDDIR>>/src/lib'
/bin/bash ../../libtool  --tag=CC   --mode=compile gcc -DPACKAGE_NAME=\"libblockdev\" -DPACKAGE_TARNAME=\"libblockdev\" -DPACKAGE_VERSION=\"2.12\" -DPACKAGE_STRING=\"libblockdev\ 2.12\" -DPACKAGE_BUGREPORT=\"vpodzime@redhat.com\" -DPACKAGE_URL=\"\" -DPACKAGE=\"libblockdev\" -DVERSION=\"2.12\" -DSTDC_HEADERS=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_MEMORY_H=1 -DHAVE_STRINGS_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_UNISTD_H=1 -DHAVE_DLFCN_H=1 -DLT_OBJDIR=\".libs/\" -DHAVE_DLFCN_H=1 -DHAVE_STRING_H=1 -DHAVE_UNISTD_H=1 -DHAVE_SYS_FCNTL_H=1 -DHAVE_SYS_IOCTL_H=1 -DHAVE_LINUX_RANDOM_H=1 -DHAVE_GLOB_H=1 -DHAVE_SYSLOG_H=1 -DHAVE_MATH_H=1 -DWITH_BD_BCACHE=/\*\*/ -DWITH_BD_BTRFS=/\*\*/ -DWITH_BD_CRYPTO=/\*\*/ -DWITH_BD_DM=/\*\*/ -DWITH_BD_LOOP=/\*\*/ -DWITH_BD_LVM=/\*\*/ -DWITH_BD_LVM_DBUS=/\*\*/ -DWITH_BD_MDRAID=/\*\*/ -DWITH_BD_MPATH=/\*\*/ -DWITH_BD_SWAP=/\*\*/ -DWITH_BD_KBD=/\*\*/ -DWITH_BD_PART=/\*\*/ -DWITH_BD_FS=/\*\*/ -I.  -I./../../include/ -Wdate-time -D_FORTIFY_SOURCE=2 -I/usr/include/glib-2.0 -I/usr/lib/s390x-linux-gnu/glib-2.0/include -I/usr/include/glib-2.0 -I/usr/lib/s390x-linux-gnu/glib-2.0/include -I/usr/include/glib-2.0 -I/usr/lib/s390x-linux-gnu/glib-2.0/include -I/usr/include/glib-2.0 -I/usr/lib/s390x-linux-gnu/glib-2.0/include -g -O2 -fdebug-prefix-map=/<<PKGBUILDDIR>>=. -fstack-protector-strong -Wformat -Werror=format-security -std=gnu99 -c -o libblockdev_la-blockdev.lo `test -f 'blockdev.c' || echo './'`blockdev.c
/bin/bash ../../libtool  --tag=CC   --mode=compile gcc -DPACKAGE_NAME=\"libblockdev\" -DPACKAGE_TARNAME=\"libblockdev\" -DPACKAGE_VERSION=\"2.12\" -DPACKAGE_STRING=\"libblockdev\ 2.12\" -DPACKAGE_BUGREPORT=\"vpodzime@redhat.com\" -DPACKAGE_URL=\"\" -DPACKAGE=\"libblockdev\" -DVERSION=\"2.12\" -DSTDC_HEADERS=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_MEMORY_H=1 -DHAVE_STRINGS_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_UNISTD_H=1 -DHAVE_DLFCN_H=1 -DLT_OBJDIR=\".libs/\" -DHAVE_DLFCN_H=1 -DHAVE_STRING_H=1 -DHAVE_UNISTD_H=1 -DHAVE_SYS_FCNTL_H=1 -DHAVE_SYS_IOCTL_H=1 -DHAVE_LINUX_RANDOM_H=1 -DHAVE_GLOB_H=1 -DHAVE_SYSLOG_H=1 -DHAVE_MATH_H=1 -DWITH_BD_BCACHE=/\*\*/ -DWITH_BD_BTRFS=/\*\*/ -DWITH_BD_CRYPTO=/\*\*/ -DWITH_BD_DM=/\*\*/ -DWITH_BD_LOOP=/\*\*/ -DWITH_BD_LVM=/\*\*/ -DWITH_BD_LVM_DBUS=/\*\*/ -DWITH_BD_MDRAID=/\*\*/ -DWITH_BD_MPATH=/\*\*/ -DWITH_BD_SWAP=/\*\*/ -DWITH_BD_KBD=/\*\*/ -DWITH_BD_PART=/\*\*/ -DWITH_BD_FS=/\*\*/ -I.  -I./../../include/ -Wdate-time -D_FORTIFY_SOURCE=2 -I/usr/include/glib-2.0 -I/usr/lib/s390x-linux-gnu/glib-2.0/include -I/usr/include/glib-2.0 -I/usr/lib/s390x-linux-gnu/glib-2.0/include -I/usr/include/glib-2.0 -I/usr/lib/s390x-linux-gnu/glib-2.0/include -I/usr/include/glib-2.0 -I/usr/lib/s390x-linux-gnu/glib-2.0/include -g -O2 -fdebug-prefix-map=/<<PKGBUILDDIR>>=. -fstack-protector-strong -Wformat -Werror=format-security -std=gnu99 -c -o libblockdev_la-plugins.lo `test -f 'plugins.c' || echo './'`plugins.c
libtool: compile:  gcc -DPACKAGE_NAME=\"libblockdev\" -DPACKAGE_TARNAME=\"libblockdev\" -DPACKAGE_VERSION=\"2.12\" "-DPACKAGE_STRING=\"libblockdev 2.12\"" -DPACKAGE_BUGREPORT=\"vpodzime@redhat.com\" -DPACKAGE_URL=\"\" -DPACKAGE=\"libblockdev\" -DVERSION=\"2.12\" -DSTDC_HEADERS=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_MEMORY_H=1 -DHAVE_STRINGS_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_UNISTD_H=1 -DHAVE_DLFCN_H=1 -DLT_OBJDIR=\".libs/\" -DHAVE_DLFCN_H=1 -DHAVE_STRING_H=1 -DHAVE_UNISTD_H=1 -DHAVE_SYS_FCNTL_H=1 -DHAVE_SYS_IOCTL_H=1 -DHAVE_LINUX_RANDOM_H=1 -DHAVE_GLOB_H=1 -DHAVE_SYSLOG_H=1 -DHAVE_MATH_H=1 "-DWITH_BD_BCACHE=/**/" "-DWITH_BD_BTRFS=/**/" "-DWITH_BD_CRYPTO=/**/" "-DWITH_BD_DM=/**/" "-DWITH_BD_LOOP=/**/" "-DWITH_BD_LVM=/**/" "-DWITH_BD_LVM_DBUS=/**/" "-DWITH_BD_MDRAID=/**/" "-DWITH_BD_MPATH=/**/" "-DWITH_BD_SWAP=/**/" "-DWITH_BD_KBD=/**/" "-DWITH_BD_PART=/**/" "-DWITH_BD_FS=/**/" -I. -I./../../include/ -Wdate-time -D_FORTIFY_SOURCE=2 -I/usr/include/glib-2.0 -I/usr/lib/s390x-linux-gnu/glib-2.0/include -I/usr/include/glib-2.0 -I/usr/lib/s390x-linux-gnu/glib-2.0/include -I/usr/include/glib-2.0 -I/usr/lib/s390x-linux-gnu/glib-2.0/include -I/usr/include/glib-2.0 -I/usr/lib/s390x-linux-gnu/glib-2.0/include -g -O2 -fdebug-prefix-map=/<<PKGBUILDDIR>>=. -fstack-protector-strong -Wformat -Werror=format-security -std=gnu99 -c blockdev.c  -fPIC -DPIC -o .libs/libblockdev_la-blockdev.o
libtool: compile:  gcc -DPACKAGE_NAME=\"libblockdev\" -DPACKAGE_TARNAME=\"libblockdev\" -DPACKAGE_VERSION=\"2.12\" "-DPACKAGE_STRING=\"libblockdev 2.12\"" -DPACKAGE_BUGREPORT=\"vpodzime@redhat.com\" -DPACKAGE_URL=\"\" -DPACKAGE=\"libblockdev\" -DVERSION=\"2.12\" -DSTDC_HEADERS=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_MEMORY_H=1 -DHAVE_STRINGS_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_UNISTD_H=1 -DHAVE_DLFCN_H=1 -DLT_OBJDIR=\".libs/\" -DHAVE_DLFCN_H=1 -DHAVE_STRING_H=1 -DHAVE_UNISTD_H=1 -DHAVE_SYS_FCNTL_H=1 -DHAVE_SYS_IOCTL_H=1 -DHAVE_LINUX_RANDOM_H=1 -DHAVE_GLOB_H=1 -DHAVE_SYSLOG_H=1 -DHAVE_MATH_H=1 "-DWITH_BD_BCACHE=/**/" "-DWITH_BD_BTRFS=/**/" "-DWITH_BD_CRYPTO=/**/" "-DWITH_BD_DM=/**/" "-DWITH_BD_LOOP=/**/" "-DWITH_BD_LVM=/**/" "-DWITH_BD_LVM_DBUS=/**/" "-DWITH_BD_MDRAID=/**/" "-DWITH_BD_MPATH=/**/" "-DWITH_BD_SWAP=/**/" "-DWITH_BD_KBD=/**/" "-DWITH_BD_PART=/**/" "-DWITH_BD_FS=/**/" -I. -I./../../include/ -Wdate-time -D_FORTIFY_SOURCE=2 -I/usr/include/glib-2.0 -I/usr/lib/s390x-linux-gnu/glib-2.0/include -I/usr/include/glib-2.0 -I/usr/lib/s390x-linux-gnu/glib-2.0/include -I/usr/include/glib-2.0 -I/usr/lib/s390x-linux-gnu/glib-2.0/include -I/usr/include/glib-2.0 -I/usr/lib/s390x-linux-gnu/glib-2.0/include -g -O2 -fdebug-prefix-map=/<<PKGBUILDDIR>>=. -fstack-protector-strong -Wformat -Werror=format-security -std=gnu99 -c plugins.c  -fPIC -DPIC -o .libs/libblockdev_la-plugins.o
In file included from blockdev.c:30:0:
plugin_apis/s390.h:7:10: fatal error: s390utils/vtoc.h: No such file or directory
 #include <s390utils/vtoc.h>
```

### PR DESCRIPTION
**Pull request by vojtechtrefny**
Added support for disabling s390 plugin and fixed some bugs found when testing the changed.

Fixes #274

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -47,10 +47,30 @@ AS_IF([test "x$found_introspection" = xyes],
 uname -p|grep s390
 on_s390=$?
```
Comment (mbiebl) [inline configure.ac]: This configure check is broken. Running ```uname -p``` on own of our s390x builder boxes yields ```unknown```
I'd propose a patch like this instead using AC_CANONICAL_HOST
```
-uname -p|grep s390
-on_s390=$?
+AC_CANONICAL_HOST
»
-AM_CONDITIONAL(ON_S390, test "$on_s390" = "0")
+case $host_cpu in
+  s390|s390x)
+    AM_CONDITIONAL(ON_S390, true)
+    ;;
+esac
```

Comment (vojtechtrefny) [reply configure.ac]: Thanks, `AC_CANONICAL_HOST` is definitely better way to do this. Fixed.

**Comment by vojtechtrefny** [comment]
@mbiebl I agree that `<s390utils/vtoc.h>` seems to be unused. I'll check with the original author of s390 plugin but I think we can safely remove this dependency.

File: configure.ac [hunk 2]
```diff
@@ -44,8 +44,11 @@ AS_IF([test "x$found_introspection" = xyes],
       [AC_SUBST(WITH_GI, 1)],
       [AC_SUBST(WITH_GI, 0)])
 
-uname -p|grep s390
-on_s390=$?
+AC_CANONICAL_HOST
+case $host_cpu in
+  s390|s390x) on_s390=0 ;;
+  *) on_s390=1 ;;
```
Comment (mbiebl) [inline configure.ac]: The use of 0 (true) and 1 (false) is a bit confusing here.
It would be more readable if you use ```on_s390=yes|no``` imho

Comment (mbiebl) [reply configure.ac]: Especially considering a line like this:
```AS_IF([test "$on_s390" = "1" -a "x$with_s390" = "xyes"],```

Comment (vojtechtrefny) [reply configure.ac]: I did want do keep the same values as before (return code of the `uname | grep`) but this is definitely better. Fixed.

**Comment by mbiebl** [comment]
Thanks for the changes. lgtm from a purely build system specific view now.

I have to admit, that I don't know what the s390 plugin actually does and this is certainly getting OT (so feel free to ignore/redirect me to a better forum): Can you give me a few pointers what functionality the s390 plugin provides and if we should enable that in the Debian package. Is there a frontend already using that plugin?

**Comment by vojtechtrefny** [comment]
> Can you give me a few pointers what functionality the s390 plugin provides and if we should enable that in the Debian package. Is there a frontend already using that plugin?

We are currently using it only in [Anaconda](https://github.com/rhinstaller/anaconda) (Fedora/RHEL installer) and [Blivet](https://github.com/storaged-project/blivet) (storage library used by Anaconda) for installing on s390 (formatting DASDs, checking some user input etc.). It isn't used by Udisks.

**Comment by mbiebl** [comment]
There is still a left-over ```#include <s390utils/vtoc.h>``` in src/lib/plugin_apis/s390.api. That one should probably be removed as well as a follow-up to a30f370

**Comment by vojtechtrefny** [comment]
> There is still a left-over #include <s390utils/vtoc.h> in src/lib/plugin_apis/s390.api.

I already removed this one in 8f9a155a9305c6f02823a321bf266d57e7dbba81

**Pull request merged**